flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -245,8 +280,38 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
             @Nonnull CheckpointStreamFactory streamFactory,
             @Nonnull CheckpointOptions checkpointOptions)
             throws Exception {
-        return keyedStateBackend.snapshot(
-                checkpointId, timestamp, streamFactory, checkpointOptions);
+        // The range to upload may overlap with the previous one(s). To reuse them, we could store
+        // the previous results either here in the backend or in the writer. However,
+        // materialization may truncate only a part of the previous result and the backend would
+        // have to split it somehow for the former option, so the latter is used.
+        lastCheckpointId = checkpointId;
+        lastUploadedFrom = materializedTo;
+        lastUploadedTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+        LOG.debug(
+                "snapshot for checkpoint {}, change range: {}..{}",
+                checkpointId,
+                lastUploadedFrom,
+                lastUploadedTo);
+        return toRunnableFuture(
+                stateChangelogWriter
+                        .persist(lastUploadedFrom)
```

Comment:
This returns the Future of StateChangelogHandle that is from 

lastUoloadedFrom till the most recent change

But "nonMaterialized" is not updated accordingly. 

Would this have problems in buildSnapshotResult

Reply:
No, `buildSnapshotResult` receives new changes as a future callback. It then combines them with `materialized` and `nonMaterialized`.
(as commented [above](https://github.com/apache/flink/pull/15420#discussion_r654955828), `nonMaterialized` only contains the state changes from the previous attempt and should be cleared after the first materialization).

Reply:
Yes (all state backend methods are assumed to be called only by the task thread).